### PR TITLE
bidirectional lane connection UI

### LINKED ISSUE #1214: bi-directional lane connection UI
now that lane connections are uni-directional, there is a need for UI to have special code for bi-directional lanes.
- Modifier-key + click creates connection in both directions. (click creates connection in both directions but modifier+click creates in one directions)
- connections should have arrow heads (we can limit this to bidirectional lanes to improve performance if rendering overlays).

Note: the current lane connector tool is confused and bugged when dealing with bi-directional lanes. It feels as though it is expecting uni-directional connections. with my uni-directional fix at the back-end (#1211) , the UI is no longer confused but rather uni-directional.

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #1214 

- Added `VectorUtil` to rotate by +-90 degree
- Added `DrawArrowHead()` to `Highlight`
- Added `IsBidirectional` to `LaneEnd` class (see code comments). this helps me decide if I need to draw arrow heads or not.
- `DrawLaneCurve()` now can draw arrow head at 2/3rd of the way.
    - When drawing lane curve, we can draw the arrow head too if necessary.
    - for backward lane curve we can draw the backward arrow head on top. 
- when shift is pressed, both forward/backward connections are created or removed.

### CODE REVIEW

**Comment by kianzarrin** [comment]
this is how it looks like. I can tweak the size
https://user-images.githubusercontent.com/26344691/160184632-27b8e45f-91bf-41ba-b356-8029e0919748.png

**Comment by kianzarrin** [comment]
image
this is also possible. Also I can make it triangle.
I can also easily play with angles.

**Comment by aubergine10** [comment]
Could the arrowhead be solid triangle in same colour as the connector line?

**Comment by kianzarrin** [comment]
@aubergine10 , I can make them triangle but not solid triangle.
Also if it is the same color as the line then it will be harder to detect.

File: TLM/TLM/UI/SubTools/LaneConnectorTool.cs [hunk 1]
```diff
@@ -1349,33 +1406,53 @@ private class LaneEnd {
...
+                    bezier: ref bezier,
+                    t: 2f / 3f,
+                    color: Color.blue,
+                    length: 0.75f,
+                    size: 0.5f, //0.03f,
+                    minY: minY,
+                    maxY: maxY,
+                    alphaBlend: true,
+                    renderLimits: underground);
+            }
```
Comment (kianzarrin) [inline TLM/TLM/UI/SubTools/LaneConnectorTool.cs:1455]: @aubergine10 you can play with this part of the code to see what you like most (Hot reload works):
DrawArrowHead() for  `>` shape
DrawArrowHead2() for triangle shape
length : determines the size of the arms
size: determines the width of the arms (the thickness of the border lines cannot be controlled, increasing width increases the filled in area inside).

**Comment by krzychu124** [comment]
> @aubergine10 , I can make them triangle but not solid triangle. Also if it is the same color as the line then it will be harder to detect.

Hmm, why not use solid triangle texture and render it rotated as quad? Check out `NetTool.RenderRoadAccessArrow`

**Comment by kianzarrin** [comment]
@krzychu124 As I said my code does has alternate to use quads. its just that its empty inside. I suspect CS quads are smaller and therefore the borders fill in the whole lane.
Since our lanes has sheath, using quad for us might get a big complicated.
I give it another try

**Comment by kianzarrin** [comment]
meanwhile that I am playing with shapes. it would be nice to review the rest of the code. it does not have to wait.

**Comment by krzychu124** [comment]
> meanwhile that I am playing with shapes. it would be nice to review the rest of the code. it does not have to wait.

The thing is you don't need to build the shape... Use texture decal and just rotate it to face correct direction similar to what you already done with arrows.

**Review by krzychu124 — APPROVED** [review]

**Comment by aubergine10** [comment]
Does the triangle need to be so large (256x256) ?

**Comment by kianzarrin** [comment]
image

**Comment by kianzarrin** [comment]
complicated UI improvements will be tracked in: https://github.com/CitiesSkylinesMods/TMPE/issues/1500

**Review by aubergine10 — APPROVED** [review]
👍

**Pull request merged**